src/engine/strat_engine/backstore/identify.rs

```diff
@@ -171,6 +172,43 @@ fn find_all_stratis_devices() -> libudev::Result<HashMap<PoolUuid, HashMap<Devic
     Ok(pool_map)
 }
 
+/// Returns the physical path of the device
+fn identify_encrypted_stratis_device(_dev: &libudev::Device) -> Option<PathBuf> {
+    None
+}
+
+/// Get the physical device path from the logical activated decrypted device
+fn logical_path_to_physical_path(_path: &Path) -> Option<PathBuf> {
+    None
+}
+
+fn find_all_closed_encrypted_stratis_devices(
+    available_stratis_devices: &HashSet<PathBuf>,
+) -> libudev::Result<HashSet<PathBuf>> {
+    let context = libudev::Context::new()?;
+
+    let opened_encrypted_devices = available_stratis_devices
+        .iter()
+        .filter_map(|path| logical_path_to_physical_path(path.as_path()))
+        .collect::<HashSet<_>>();
+
+    let mut enumerator = block_enumerator(&context)?;
+    enumerator.match_property("ID_FS_TYPE", "crypto_LUKS")?;
+
+    let encrypted_devices: Vec<_> = enumerator.scan_devices()?.collect();
+
+    let closed_encrypted_device_map = encrypted_devices
+        .iter()
+        .filter_map(|dev| identify_encrypted_stratis_device(dev))
```

Comment:
Unless ```identify_encrypted_stratis_device``` does something that I don't understand you can use existing udev method ```devnode``` here.

Reply:
Per stand up, I think this may be a misunderstanding as I realize my comments in the PR are a bit vague.

The algorithm takes the following steps:
1. It takes all available Stratis devices as an input. This could be unencrypted devices or encrypted devices that have already been opened.
2. It filters out all unencrypted devices using `logical_path_to_physical_path` which will return `None` for unencrypted devices and the path of the physical device by using the libcryptsetup bindings for all logical `/dev/mapper` devices corresponding to an encrypted physical device that has been opened.
3. It searches again, this time for all encrypted physical volumes using the `crypto_LUKS` udev type.
4. It uses `identify_encrypted_stratis_device` to probe for our Stratis token to determine whether or not the LUKS2 volume we are looking at is indeed a Stratis encrypted volume. If it is not, it returns `None`. If it is, it returns the same thing as `devnode` as you suggested. My guess is that I will end up using `devnode` internally in `identify_encrypted_stratis_device` but the two are not currently interchangeable as it stands in my design.
5. Filter out all of the devices that have already been determined to be opened in step 2 by checking if `opened_encrypted_devices` contains the physical path that we mapped to in step 4 using `identify_encrypted_stratis_device`. If it is not already opened, pass it along to the next step.
6. Return all unopened devices from the function to be opened [here](https://github.com/stratis-storage/stratisd/pull/1834/files#diff-5434ab93d532c0d2d6f44f725907f5e6R268).

Does this make a little more sense? There may be room for improvement still but hopefully this explains it a little bit better.

Reply:
> Per stand up, I think this may be a misunderstanding as I realize my comments in the PR are a bit vague.
> 
> The algorithm takes the following steps:
> 
> 1. It takes all available Stratis devices as an input. This could be unencrypted devices or encrypted devices that have already been opened.
> 2. It filters out all unencrypted devices using `logical_path_to_physical_path` which will return `None` for unencrypted devices and the path of the physical device by using the libcryptsetup bindings for all logical `/dev/mapper` devices corresponding to an encrypted physical device that has been opened.
> 3. It searches again, this time for all encrypted physical volumes using the `crypto_LUKS` udev type.
> 4. It uses `identify_encrypted_stratis_device` to probe for our Stratis token to determine whether or not the LUKS2 volume we are looking at is indeed a Stratis encrypted volume. If it is not, it returns `None`. If it is, it returns the same thing as `devnode` as you suggested. My guess is that I will end up using `devnode` internally in `identify_encrypted_stratis_device` but the two are not currently interchangeable as it stands in my design.
> 5. Filter out all of the devices that have already been determined to be opened in step 2 by checking if `opened_encrypted_devices` contains the physical path that we mapped to in step 4 using `identify_encrypted_stratis_device`. If it is not already opened, pass it along to the next step.
> 6. Return all unopened devices from the function to be opened [here](https://github.com/stratis-storage/stratisd/pull/1834/files#diff-5434ab93d532c0d2d6f44f725907f5e6R268).
> 
> Does this make a little more sense? There may be room for improvement still but hopefully this explains it a little bit better.

Yes. That is quite clear. You might as well close this now; I've adopted it into https://github.com/stratis-storage/stratisd/pull/1814 and will try to proceed from there, probably stubbing out some code about the event-based portion of the device discovery in the same spirit.